STIPS/STIP-002.md

```diff
@@ -0,0 +1,150 @@
+    - Also, issuance/redemption generally would not lead to overcollaterlization
+        - For it to happen, we would need the special case of a token which rewards holders on a transfer
+        - we can handle that case using Airdrop module
+
+3. How to handle `newBalance < existingBalance.add(_quantity))`?
+    - This is a possible case for aTokens, since `(quantity/index) * index` can be less than `quantity`.
+    - We have seen this case once during testing.
+
+
+## Checkpoint 1
+
+**Reviewer**:
+
+## Proposed Architecture Changes
+
+Pseudocode of option1 :
+```javascript
+
+library Invoke {
+    ... // other functions from Invoke library
+    function strictInvokeTransfer(ISetToken _setToken, address _token, address _to, uint256 _quantity) internal {
+        ../// existing implementations
+    }
+    function invokeTransferWithCollaterlizationChecks(ISetToken _setToken, address _token, address _to, uint256 _quantity) internal {
+        if (_quantity > 0) {
+            // Retrieve current balance of token for the SetToken
+            uint256 existingBalance = IERC20(_token).balanceOf(address(_setToken));
+
+            Invoke.invokeTransfer(_setToken, _token, _to, _quantity);
+
+            // Get new balance of transferred token for SetToken
+            uint256 newBalance = IERC20(_token).balanceOf(address(_setToken));
+
+            // prevent undercollateralization
+            require(newBalance >= existingBalance.sub(_quantity)), "Invalid post transfer balance");
+        }
+    }
+}
+
+
+// Here we can either add a new library with a new transferFrom function
+// or add a new function in the ExplicitERC20 library.
+library ExplicitERC20V2 {
+    function transferFromWithCollaterlizationChecks(IERC20 _token, address _from, address _to, uint256 _quantity) internal {
+        if (_quantity > 0) {
+            uint256 existingBalance = _token.balanceOf(_to);
+            SafeERC20.safeTransferFrom(_token, _from, _to, _quantity);
+            uint256 newBalance = _token.balanceOf(_to);
+
+            // prevent undercollateralization
+            require(
+                newBalance >= existingBalance.add(_quantity)), "Invalid post transfer balance"
+            );
+        }
+    }
+}
+
+// Since we use ModuleBase#transferFrom in our issuance modules we would also need to add a new function in the ModuleBase contract which uses the new ExplicitERC20#transferFromWithCollaterlizationChecks
+// Or we can skip calling ModuleBase#transferFrom in our new issuance modules and directly call ExplicitERC20#transferFromWithCollaterlizationChecks
+abstract contract ModuleBase is IModule {
+    ...//other existing functions
+    function transferFromWithCollaterlizationChecks(IERC20 _token, address _from, address _to, uint256 _quantity) internal {
+        ExplicitERC20.transferFromWithCollaterlizationChecks(_token, _from, _to, _quantity);
+    }
+}
+```
+Next, when deploying the new issuance modules we link the modified new libraries to them.
```

Comment:
I think we'll need to change the actual Module code as well since we need to make sure we call the correct library code. That module should be named differently to make it clear that the security assumptions are slightly different. Luckily I think we should just be able to inherit the DebtIssuanceModule and `override` `_resolveEquityPositions` and `_resolveDebtPositions`.